### DISCUSSION TITLE
Fixing mod's language mistakes and allowing translation to different languages.

### DISCUSSION THREAD

**Discussion by BreadDain** [IHateMyKite/UnforgivingDevices#22] [Ideas]
xTranslator https://www.nexusmods.com/skyrimspecialedition/mods/134?tab=description is a very useful tool to edit plugin messages or record names to different language, possibly utilizing autotranslate functionality like GoogleTranslate or even DeepL api, which has a remarkable quality.

However, there are certain strings spread across scripts, they can be found with `Print.*?".*?".*$` or `debug.notification.*?".*?".*$` regExps, but less capable users wouldn't be able to perform such trick since they usually are not very well used to programming in general. Solution here would be using translation.txt files added to mod and making all script functions fetch those lines from that file depending on localization. xTranslator is also able to process those, so full translation process would be simple enough to perform for any user, as simple as device patching procedure.

**Comment by BreadDain** [comment]
Regexp to find all? text output in the code, Notepad or any .net standard: `^.*?(SetInfoText\(.*?".*".*?\))|(notification\(.*?".*".*?\))|(messagebox\(.*?".*".*?\))*(Print\(.*?".*".*?\))|(showMessageBox\(.*?".*".*?\)).*?$`

**Comment by BreadDain** [comment]
> You should be able to just call Debug.Notification("$test") in a script as long as SKSE is enabled when you run Skyrim. I tested it and it works. One thing you might have missed is to set the encoding of the translation file to UCS-2 Little Endian (or UCS-2 LE), which is a critical step.

Mix strings with functions: https://www.loverslab.com/topic/108918-translation-of-scripts/?do=findComment&comment=2393282

> Since Notepad++ 6.0 it is actually possible without scripts using the following technique:
> 
>     Find: Text_1 and Replace with: Text_A
> 
>     Find: Text2 and Replace with: TextB
> 
> Find: (Text_1)|(Text2)
> 
> Replace: (?1Text_A)(?2TextB)
> 
> The general syntax is:
> 
> Find: (FindA)|(FindB)|(FindC)...
> 
> Replace: (?1ReplaceA)(?2ReplaceB)(?3ReplaceC)...
> 
> ?1 refers to the first captured phrase and so on.
> 
> I got this from the following answer: https://stackoverflow.com/a/16104946/287101

but on another hand there is [a mod](https://www.nexusmods.com/skyrimspecialedition/mods/22603?tab=description) that tells that is supposed to enable nesting... doesn't work by default? need to test on practice

**Comment by IHateMyKite** [comment]
Did try the $ approach. It looks promising, but there are some things that need to be tested first (like messageboxes, which play big role in UD). I will add the translation file in the next update, and then it can be slowly worked on.
